From 41267caca2290a7a1e7556e717f8b408c57f08f4 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Mon, 26 Apr 2021 08:28:50 -0500
Subject: [PATCH 085/196] Refactor GetBlockSubsidy to depend on nBits instead
 of nHeight (#155)

This will allow us to build a coin where the block reward depends on difficulty.
---
 src/bench/duplicate_inputs.cpp |  2 +-
 src/miner.cpp                  | 13 ++++++------
 src/rpc/blockchain.cpp         |  2 +-
 src/test/validation_tests.cpp  | 36 ++++++++++++++++++++++++++++------
 src/validation.cpp             |  5 +++--
 src/validation.h               |  3 ++-
 6 files changed, 44 insertions(+), 17 deletions(-)

diff --git a/src/bench/duplicate_inputs.cpp b/src/bench/duplicate_inputs.cpp
index 17be6eeed..f3e86b125 100644
--- a/src/bench/duplicate_inputs.cpp
+++ b/src/bench/duplicate_inputs.cpp
@@ -46,7 +46,7 @@ static void DuplicateInputs(benchmark::Bench &bench) {
     coinbaseTx.vin[0].prevout = COutPoint();
     coinbaseTx.vout.resize(1);
     coinbaseTx.vout[0].scriptPubKey = SCRIPT_PUB;
-    coinbaseTx.vout[0].nValue = GetBlockSubsidy(nHeight, consensusParams);
+    coinbaseTx.vout[0].nValue = GetBlockSubsidy(block.nBits, consensusParams);
     coinbaseTx.vin[0].scriptSig = CScript() << COINBASE_PREFIX
                                             << nHeight
                                             << OP_0;
diff --git a/src/miner.cpp b/src/miner.cpp
index 747c85d80..583d7a227 100644
--- a/src/miner.cpp
+++ b/src/miner.cpp
@@ -148,7 +148,12 @@ BlockAssembler::CreateNewBlock(const CScript &scriptPubKeyIn) {
         pblock->nVersion = gArgs.GetArg("-blockversion", pblock->nVersion);
     }
 
+    // Fill in header.
+    pblock->hashPrevBlock = pindexPrev->GetBlockHash();
     pblock->nTime = GetAdjustedTime();
+    UpdateTime(pblock, chainParams, pindexPrev);
+    pblock->nBits = GetNextWorkRequired(pindexPrev, pblock, chainParams);
+    pblock->nNonce = 0;
     nLockTimeCutoff = pindexPrev->GetMedianTimePast();
 
     int nPackagesSelected = 0;
@@ -181,7 +186,8 @@ BlockAssembler::CreateNewBlock(const CScript &scriptPubKeyIn) {
     coinbaseTx.vout.resize(1);
     coinbaseTx.vout[0].scriptPubKey = scriptPubKeyIn;
     coinbaseTx.vout[0].nValue =
-        amountFeeReward + GetBlockSubsidy(nHeight, consensusParams);
+        amountFeeReward +
+        GetBlockSubsidy(pblock->nBits, consensusParams);
     coinbaseTx.vin[0].scriptSig = CScript()
                                   << COINBASE_PREFIX << nHeight << OP_0;
 
@@ -209,11 +215,6 @@ BlockAssembler::CreateNewBlock(const CScript &scriptPubKeyIn) {
     LogPrintf("CreateNewBlock(): total size: %u txs: %u fees: %ld sigops %d\n",
               nSerializeSize, nBlockTx, nFees, nBlockSigOps);
 
-    // Fill in header.
-    pblock->hashPrevBlock = pindexPrev->GetBlockHash();
-    UpdateTime(pblock, chainParams, pindexPrev);
-    pblock->nBits = GetNextWorkRequired(pindexPrev, pblock, chainParams);
-    pblock->nNonce = 0;
     pblocktemplate->entries[0].sigOpCount = 0;
 
     BlockValidationState state;
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index 211e3365a..ffa2c2b91 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -2405,7 +2405,7 @@ static UniValue getblockstats(const Config &config,
     ret_all.pushKV("mintxsize", mintxsize == blockMaxSize ? 0 : mintxsize);
     ret_all.pushKV("outs", outputs);
     ret_all.pushKV("subsidy", ValueFromAmount(GetBlockSubsidy(
-                                  pindex->nHeight, Params().GetConsensus())));
+                                  pindex->nBits, Params().GetConsensus())));
     ret_all.pushKV("time", pindex->GetBlockTime());
     ret_all.pushKV("total_out", ValueFromAmount(total_out));
     ret_all.pushKV("total_size", total_size);
diff --git a/src/test/validation_tests.cpp b/src/test/validation_tests.cpp
index 3e608c0ce..16b7ea5d8 100644
--- a/src/test/validation_tests.cpp
+++ b/src/test/validation_tests.cpp
@@ -30,12 +30,36 @@ BOOST_FIXTURE_TEST_SUITE(validation_tests, TestingSetup)
  * Make sure the block subsidy is really constant.
  */
 static void TestBlockSubsidyConstant(const Consensus::Params &consensusParams) {
-    BOOST_CHECK_EQUAL(SUBSIDY, GetBlockSubsidy(0, consensusParams));
-    BOOST_CHECK_EQUAL(SUBSIDY, GetBlockSubsidy(1, consensusParams));
-    BOOST_CHECK_EQUAL(SUBSIDY, GetBlockSubsidy(100'000, consensusParams));
-    BOOST_CHECK_EQUAL(SUBSIDY, GetBlockSubsidy(210'000, consensusParams));
-    BOOST_CHECK_EQUAL(SUBSIDY, GetBlockSubsidy(210'001, consensusParams));
-    BOOST_CHECK_EQUAL(SUBSIDY, GetBlockSubsidy(0x7fff'ffff, consensusParams));
+    BOOST_CHECK_EQUAL(SUBSIDY, GetBlockSubsidy(0x1d00ffff, consensusParams));
+    BOOST_CHECK_EQUAL(SUBSIDY, GetBlockSubsidy(0x1d008000, consensusParams));
+    BOOST_CHECK_EQUAL(SUBSIDY, GetBlockSubsidy(0x1e7fffff, consensusParams));
+    BOOST_CHECK_EQUAL(
+        SUBSIDY,
+        GetBlockSubsidy(0x1c2a1115, consensusParams)); // block 50000 on BTC
+    BOOST_CHECK_EQUAL(
+        SUBSIDY,
+        GetBlockSubsidy(0x1b04864c, consensusParams)); // block 100000 on BTC
+    BOOST_CHECK_EQUAL(
+        SUBSIDY,
+        GetBlockSubsidy(0x1a05db8b, consensusParams)); // block 200000 on BTC
+    BOOST_CHECK_EQUAL(
+        SUBSIDY,
+        GetBlockSubsidy(0x1900896c, consensusParams)); // block 300000 on BTC
+    BOOST_CHECK_EQUAL(
+        SUBSIDY,
+        GetBlockSubsidy(0x1806b99f, consensusParams)); // block 400000 on BTC
+    BOOST_CHECK_EQUAL(
+        SUBSIDY,
+        GetBlockSubsidy(0x18009645, consensusParams)); // block 500000 on BTC
+    BOOST_CHECK_EQUAL(
+        SUBSIDY,
+        GetBlockSubsidy(0x1715a35c, consensusParams)); // block 600000 on BTC
+    BOOST_CHECK_EQUAL(
+        SUBSIDY,
+        GetBlockSubsidy(0x170bef93, consensusParams)); // block 680603 on BTC
+    BOOST_CHECK_EQUAL(SUBSIDY, GetBlockSubsidy(0x1000ffff, consensusParams));
+    BOOST_CHECK_EQUAL(SUBSIDY, GetBlockSubsidy(0x0500ffff, consensusParams));
+    BOOST_CHECK_EQUAL(SUBSIDY, GetBlockSubsidy(0x0300ffff, consensusParams));
 }
 
 static void TestBlockSubsidyConstant(int nSubsidyHalvingInterval) {
diff --git a/src/validation.cpp b/src/validation.cpp
index 342c4220a..f25e9a15d 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -860,7 +860,8 @@ static bool WriteBlockToDisk(const CBlock &block, FlatFilePos &pos,
     return true;
 }
 
-Amount GetBlockSubsidy(int nHeight, const Consensus::Params &consensusParams) {
+Amount GetBlockSubsidy(uint32_t nBits,
+                       const Consensus::Params &consensusParams) {
     return SUBSIDY;
 }
 
@@ -1847,7 +1848,7 @@ bool CChainState::ConnectBlock(const CBlock &block, BlockValidationState &state,
     // 50% of fees get burned.
     Amount amountFeeReward = nFees / 2;
     Amount blockReward =
-        amountFeeReward + GetBlockSubsidy(pindex->nHeight, consensusParams);
+        amountFeeReward + GetBlockSubsidy(block.nBits, consensusParams);
     if (block.vtx[0]->GetValueOut() > blockReward) {
         LogPrintf("ERROR: ConnectBlock(): coinbase pays too much (actual=%d vs "
                   "limit=%d)\n",
diff --git a/src/validation.h b/src/validation.h
index 3b63035c9..21e535570 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -249,7 +249,8 @@ bool GetTransaction(const TxId &txid, CTransactionRef &txOut,
 bool ActivateBestChain(
     const Config &config, BlockValidationState &state,
     std::shared_ptr<const CBlock> pblock = std::shared_ptr<const CBlock>());
-Amount GetBlockSubsidy(int nHeight, const Consensus::Params &consensusParams);
+Amount GetBlockSubsidy(uint32_t nBits,
+                       const Consensus::Params &consensusParams);
 
 /**
  * Guess verification progress (as a fraction between 0.0=genesis and
-- 
2.39.5 (Apple Git-154)

